tests/jobs/localisation/example_case/scripts/common_functions.py

```diff
@@ -0,0 +1,604 @@
+        try:
+            seed_value = int(lines[r_number - 1])
+        except IndexError as exc:
+            raise IOError("Seed value not found for realization {r_number}  ") from exc
+        except ValueError as exc:
+            raise IOError(
+                "Invalid seed value in file for realization{r_number}"
+            ) from exc
+    return seed_value
+
+
+def upscaling(field_values, write_field=True, iteration=0):
+    """
+    Calculate upscaled values and optionally write upscaled values to file.
+    Return upscaled values
+    """
+    response_function_name = settings.response.response_function
+
+    NX, NY, NZ = settings.response.grid_dimension
+    calculate_all = settings.response.calculate_all_cells
+    coarse_cell_index_list = settings.observation.selected_grid_cells
+    upscaled_values = np.zeros((NX, NY, NZ), dtype=np.float32, order="F")
+    upscaled_values[:, :, :] = -999
+
+    if response_function_name == "average":
+        upscaled_values = upscale_average(
+            field_values, coarse_cell_index_list, upscaled_values, use_all=calculate_all
+        )
+
+    if write_field:
+        upscaled_file_name = settings.response.upscaled_file_name
+        if iteration == 0:
+            upscaled_file_name = "init_files/" + upscaled_file_name
+        write_upscaled_field(upscaled_values, upscaled_file_name)
+
+    return upscaled_values
+
+
+def write_upscaled_field(
+    upscaled_values, upscaled_file_name, selected_cell_index_list=None
+):
+    nx, ny, nz = upscaled_values.shape
+    field_name = "Upscaled"
+
+    field_object = xtgeo.grid3d.GridProperty(
+        ncol=nx,
+        nrow=ny,
+        nlay=nz,
+        values=upscaled_values,
+        discrete=False,
+        name=field_name,
+    )
+
+    print(f"Write upscaled field file: {upscaled_file_name}  ")
+    field_object.to_file(upscaled_file_name, fformat="roff")
+    if selected_cell_index_list is not None:
+        selected_upscaled_values = np.zeros((nx, ny, nz), dtype=np.float32, order="F")
+        selected_upscaled_values[:, :, :] = -1
+        nobs = get_nobs()
+        for obs_number in range(nobs):
+            (Iindx, Jindx, Kindx) = get_cell_indices(
+                obs_number, nobs, selected_cell_index_list
+            )
+            selected_upscaled_values[Iindx, Jindx, Kindx] = upscaled_values[
+                Iindx, Jindx, Kindx
+            ]
+
+        field_name_selected = field_name + "_conditioned_cells"
+        file_name_selected = "init_files/" + field_name_selected + ".roff"
+        cond_field_object = xtgeo.grid3d.GridProperty(
+            ncol=nx,
+            nrow=ny,
+            nlay=nz,
+            values=selected_upscaled_values,
+            discrete=False,
+            name=field_name_selected,
+        )
+        print(f"Write conditioned cell values as field: {file_name_selected}")
+        cond_field_object.to_file(file_name_selected, fformat="roff")
+
+    return field_object
+
+
+def upscale_average(
```

Comment:
This function is quite complex and should really have a test of its own. Could we leverage some third-party library that already has upscaling implemented? I did a search for upscaling in `xtgeo` but came up short.

If we really need to write this our selves, we could perhaps at least try and leverage `np.mean`. Something like this:

```python
import numpy as np

def upscale_average(field_values, coarse_cell_index_list, upscaled_values, use_all=False):
    nx, ny, nz = field_values.shape
    NX, NY, NZ = upscaled_values.shape

    print(f"Number of fine scale grid cells:   (nx,ny,nz): ({nx},{ny},{nz})")
    print(f"Number of coarse scale grid cells: (NX,NY,NZ): ({NX},{NY},{NZ})  ")

    mx, my, mz = nx // NX, ny // NY, nz // NZ
    print(
        "Number of fine scale grid cells per coarse grid cell: "
        f"(mx,my,mz): ({mx},{my},{mz})"
    )

    if use_all:
        print("Calculate upscaled values for all grid cells")
        reshaped_field = field_values.reshape(NX, mx, NY, my, NZ, mz)
        avg_values = np.mean(reshaped_field, axis=(1, 3, 5))
        np.copyto(upscaled_values, avg_values)
    else:
        print("Calculate upscaled values for selected grid cells")
        for Iindx, Jindx, Kindx in coarse_cell_index_list:
            Iindx, Jindx, Kindx = Iindx - 1, Jindx - 1, Kindx - 1  # adjust for 1-based index

            i_slice = slice(Iindx * mx, (Iindx + 1) * mx)
            j_slice = slice(Jindx * my, (Jindx + 1) * my)
            k_slice = slice(Kindx * mz, (Kindx + 1) * mz)

            avg_value = np.mean(field_values[i_slice, j_slice, k_slice])
            upscaled_values[Iindx, Jindx, Kindx] = avg_value

    return upscaled_values

```

Reply:
I have done some update using numpy slice and mean and simplified the function